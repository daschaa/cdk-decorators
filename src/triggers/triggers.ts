import { Stack } from 'aws-cdk-lib';
import * as triggers from 'aws-cdk-lib/triggers';

/**
 * Decorator to run a lambda trigger after the stack deployment.
 *
 * @param triggerProps parameters for the trigger
 * @example
 * ```ts
 * @TriggerAfter({
 *   code: Code.fromInline(`
 *       exports.handler = async (event) => {
 *         console.log('event: ', event)
 *       };
 *     `),
 *   handler: 'index.handler',
 *   runtime: Runtime.NODEJS_14_X,
 * })
 * class TestStack extends Stack {
 *   constructor(scope: any, id: string) {
 *     super(scope, id);
 *   }
 * }
 * ```
 */
export function TriggerAfter(triggerProps: triggers.TriggerFunctionProps): any {
  return function<T extends { new (...args: any[]): Stack }>(ctor: T) {
    return class extends ctor {
      constructor(...args: any[]) {
        super(...args);
        const trigger = new triggers.TriggerFunction(this, 'TriggerAfter', triggerProps);
        trigger.executeAfter(this);
      }
    };
  };
}
