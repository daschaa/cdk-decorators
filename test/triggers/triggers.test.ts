import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { TriggerAfter } from '../../src/triggers/triggers';

describe('TriggerAfter', () => {

  @TriggerAfter({
    code: Code.fromInline(`
      exports.handler = async (event) => {
        console.log('event: ', event)
      };
    `),
    handler: 'index.handler',
    runtime: Runtime.NODEJS_14_X,
  })
  class TestStack extends Stack {
    constructor(scope: any, id: string) {
      super(scope, id);
      new Bucket(this, 'TestBucket');
    }
  }

  test('creates trigger for stack', () => {
    const app = new App();
    const stack = new TestStack(app, 'TestStack');
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
