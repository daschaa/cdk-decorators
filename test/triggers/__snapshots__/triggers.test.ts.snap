// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerAfter creates trigger for stack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": {
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "fd4ae44810153bccde8a320d4b4fc240cedcbef58d7cf322707ea2a46f7c0150.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "TriggerAfter2216DD10",
                              "Arn",
                            ],
                          },
                          ":*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestBucket560B80BC": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TriggerAfter2216DD10": {
      "DependsOn": [
        "TriggerAfterServiceRole274B9506",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "
      exports.handler = async (event) => {
        console.log('event: ', event)
      };
    ",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TriggerAfterServiceRole274B9506",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TriggerAfterCurrentVersionA734A45098bc03cc66e4ce19e66e864fe987b728": {
      "Properties": {
        "FunctionName": {
          "Ref": "TriggerAfter2216DD10",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "TriggerAfterServiceRole274B9506": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TriggerAfterTrigger375D8B0D": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
        "TestBucket560B80BC",
        "TriggerAfterCurrentVersionA734A45098bc03cc66e4ce19e66e864fe987b728",
        "TriggerAfter2216DD10",
        "TriggerAfterServiceRole274B9506",
      ],
      "Properties": {
        "HandlerArn": {
          "Ref": "TriggerAfterCurrentVersionA734A45098bc03cc66e4ce19e66e864fe987b728",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
